### PR TITLE
Creating a add_mylist methods

### PR DESCRIPTION
**Pull request by soramugi**
I tried to make method to add a video to My List

### CODE REVIEW

File: lib/niconico/api.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8 -*-
```
Comment (sorah) [inline lib/niconico/api.rb]: I guess this is unnecessary

File: lib/niconico/api.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        }
+      )
+    end
+
+    private
+    def post path, params
+      uri = URI.join(Niconico::URL[:top], path)
+      page = @agent.post(uri, params)
+      json = JSON.parse(page.body)
+      raise ApiParseError, json unless json['status'] == 'ok'
```
Comment (sorah) [inline lib/niconico/api.rb]: This is not a _parse_ error. Consider renaming exception class

File: lib/niconico.rb [hunk 3]
```diff
@@ -9,7 +9,8 @@ class Niconico
     top: 'http://www.nicovideo.jp/',
     login: 'https://secure.nicovideo.jp/secure/login?site=niconico',
     watch: 'http://www.nicovideo.jp/watch/',
-    getflv: 'http://flapi.nicovideo.jp/api/getflv'
+    getflv: 'http://flapi.nicovideo.jp/api/getflv',
+    mymylist: 'http://www.nicovideo.jp/my/mylist'
```
Comment (sorah) [inline lib/niconico.rb]: I prefer `my_mylist`

File: lib/niconico/api.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    def initialize(parent, token=nil)
+      @parent = parent
+      @agent = parent.agent
+      @token = token || get_token
+    end
+
+    def get_token
+      page = @agent.get(Niconico::URL[:mymylist])
+      if page.search("script").map(&:inner_text).find{|x| /\tNicoAPI\.token/ =~ x }.match(/\tNicoAPI\.token = "(.+)";\n/)
```
Comment (sorah) [inline lib/niconico/api.rb]: `.find{|x| ... }` can be replaced with Enumerable#grep

File: lib/niconico/api.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        @token = $1
+      else
+        raise ApiParseError, 'token can not be acquired'
```
Comment (sorah) [inline lib/niconico/api.rb]: At here we're not using API, consider creating another exception class (`AcquiringTokenError` ?)

Comment (sorah) [inline lib/niconico/api.rb]: I prefer `Couldn't find a token` as an error message

File: lib/niconico/api.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+      end
+    end
+
+    def mylist_add group_id, item_type, item_id, description=''
```
Comment (sorah) [inline lib/niconico/api.rb]: wrap parameters with `()`

File: lib/niconico/api.rb [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+          item_type: item_type,
+          item_id: item_id,
+          description: description,
+          token: @token
+        }
+      )
+    end
+
+    private
+    def post path, params
```
Comment (sorah) [inline lib/niconico/api.rb]: wrap parameters with `()`

File: lib/niconico/api.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+          # video: 0 seiga: 5
+          item_type: item_type,
+          item_id: item_id,
+          description: description,
+          token: @token
+        }
+      )
+    end
+
+    private
```
Comment (sorah) [inline lib/niconico/api.rb]: insert empty line below here

File: lib/niconico/api.rb [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        raise ApiParseError, 'token can not be acquired'
+      end
+    end
+
+    def mylist_add group_id, item_type, item_id, description=''
+      !!post(
+        '/api/mylist/add',
+        {
+          group_id: group_id,
+          # video: 0 seiga: 5
```
Comment (sorah) [inline lib/niconico/api.rb]: 1. Make them as const `API::MYLIST_ITEM_TYPES = {video: 0, seiga: 5}`
2. `item_type: API::MYLIST_ITEM_TYPES[item_type]`
3. `api.mylist_add(mylist_id, :video, @id, description)`

File: lib/niconico/api.rb [hunk 10]
```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8 -*-
+require 'json'
+
+class Niconico
+  class API
+    class ApiParseError < Exception; end
+
+    def initialize(parent, token=nil)
+      @parent = parent
+      @agent = parent.agent
```
Comment (sorah) [inline lib/niconico/api.rb]: I prefer using helper method like `def agent; @parent.agent; end`

File: lib/niconico/api.rb [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+      @token = token || get_token
```
Comment (sorah) [inline lib/niconico/api.rb]: defer acquiring token: `def token; @token ||= token || get_token; end`

File: lib/niconico/video.rb [hunk 12]
```diff
@@ -118,6 +118,11 @@ def video_cookie_jar_file
       end
     end
 
+    def add_mylist(mylist_id, description='')
```
Comment (sorah) [inline lib/niconico/video.rb]: add *to* mylist?

File: lib/niconico/api.rb [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    end
+
+    def get_token
+      page = @agent.get(Niconico::URL[:mymylist])
+      if page.search("script").map(&:inner_text).find{|x| /\tNicoAPI\.token/ =~ x }.match(/\tNicoAPI\.token = "(.+)";\n/)
+        @token = $1
```
Comment (sorah) [inline lib/niconico/api.rb]: consider using MatchData that returned from String#match instead of `$1`

**Comment by sorah** [comment]
IMO I write commit message as the following for this case:

```
Add Video#add_to_mylist

- API: class for NicoAPI
```

**Comment by soramugi** [comment]
Please check :satisfied:

File: lib/niconico/nico_api.rb [hunk 14]
```diff
@@ -0,0 +1,52 @@
+require 'json'
+
+class Niconico
+  class NicoAPI
+    class ApiStatusError < Exception; end
+    class FindTokenError < Exception; end
```
Comment (sorah) [inline lib/niconico/nico_api.rb]: FindTokenError -> AcquiringTokenError

ApiStatusError -> ApiError

**Comment by sorah** [comment]
Bit closer! Fix my additional comments then squash the changes. :sparkles:

File: lib/niconico/video.rb [hunk 15]
```diff
@@ -118,6 +118,11 @@ def video_cookie_jar_file
       end
     end
 
+    def add_to_mylist(mylist_id, description='')
+      api = Niconico::NicoAPI.new(@parent)
```
Comment (sorah) [inline lib/niconico/video.rb]: Holding NicoAPI object in Niconico object is better, it makes api token reusable

**Comment by soramugi** [comment]
Have been fixed :blush:

File: lib/niconico/video.rb [hunk 16]
```diff
@@ -118,6 +118,11 @@ def video_cookie_jar_file
       end
     end
 
+    def add_to_mylist(mylist_id, description='')
+      api = NicoAPI.new(@parent)
```
Comment (sorah) [inline lib/niconico/video.rb]: I meant to add Niconico#nico_api. This can't reuse API token :/

**Comment by soramugi** [comment]
And like when you move the page change the value of the api token.
token is always 're not a little self-confidence whether may be the same :confused:

**Comment by sorah** [comment]
なんとなく意味は分かったけど解釈しきてている自信がないので日本語でお願いします。

**Comment by sorah** [comment]
235e1a7c3d4ed4444e22b428dd7135bd6c3bb3f4 looks good

Trying to re-acquire token when token has expired is better, but it's future work.

**Comment by sorah** [comment]
Is a token expires too shortly?

**Comment by soramugi** [comment]
はい
tokenの値がページ移動すると変わるので、保持したままだとページ移動挟んでから再度実行して正常動作するかちょっと自信ないです。

**Comment by sorah** [comment]
試してみましたが速攻で expire する事はなさそうです。たぶん expire はあるしそれっぽいエラーコードも見当ったので、まあそれが出たら再取得するとかするとよさそうですが、とりあえずこのパッチについては LGTM です :sparkles:

http://uni.res.nimg.jp/js/nicoapi.js?0.1.1

----

`NicoAPI.token` changes every access, but tokens don't expire too shortly. For future work, we should do re-acquire when "INVALID_TOKEN" or "EXPIRED_TOKEN" error appears.

Anyway this patch LGTM. Merging.

**Comment by soramugi** [comment]
ありがとうございます!!!:joy:

**Pull request merged**